lib/lints/avoid_unnecessary_type_assertions/avoid_unnecessary_type_assertions_rule.dart

```diff
@@ -0,0 +1,183 @@
+      }
+    });
+
+    context.registry.addMethodInvocation((node) {
+      if (_isUnnecessaryWhereType(node)) {
+        reporter.reportErrorForNode(_unnecessaryWhereTypeCode, node);
+      }
+    });
+  }
+
+  @override
+  List<Fix> getFixes() => [_UnnecessaryTypeAssertionsFix()];
+
+  bool _isUnnecessaryIsExpression(IsExpression node) {
+    final objectType = node.expression.staticType;
+    final castedType = node.type.type;
+
+    if (node.notOperator != null) {
+      if (objectType != null &&
+          objectType is! TypeParameterType &&
+          objectType is! DynamicType &&
+          !objectType.isDartCoreObject) {
+        return _isUnnecessaryTypeCheck(
+          objectType,
+          castedType,
+          isReversed: true,
+        );
+      }
+    }
+
+    return _isUnnecessaryTypeCheck(objectType, castedType);
+  }
+
+  bool _isUnnecessaryWhereType(MethodInvocation node) {
+    const whereTypeMethodName = 'whereType';
+
+    final targetType = node.target?.staticType;
+
+    if (node.methodName.name == whereTypeMethodName &&
+        isIterableOrSubclass(node.realTarget?.staticType) &&
+        targetType is ParameterizedType) {
+      final isTargetTypeHasGeneric = targetType.typeArguments.isNotEmpty;
+      final arguments = node.typeArguments?.arguments;
+      final isWhereTypeHasGeneric = arguments?.isNotEmpty ?? false;
+
+      return isTargetTypeHasGeneric &&
+          isWhereTypeHasGeneric &&
+          _isUnnecessaryTypeCheck(
+            targetType.typeArguments.first,
+            arguments?.first.type,
+          );
+    }
+    return false;
+  }
+
+  bool _isUnnecessaryTypeCheck(
+    DartType? objectType,
+    DartType? castedType, {
+    bool isReversed = false,
+  }) {
+    if (objectType == null || castedType == null) {
+      return false;
+    }
+
+    if (_isNullableCompatibility(objectType, castedType)) {
+      return false;
+    }
+
+    final objectCastedType =
+        _castedTypeInObjectTypeHierarchy(objectType, castedType);
+
+    if (objectCastedType == null) {
+      return isReversed;
+    }
+
+    if (!_isGenerics(objectCastedType, castedType)) {
+      return false;
+    }
+
+    return !isReversed;
+  }
+
+  bool _isNullableCompatibility(DartType objectType, DartType castedType) {
+    final isObjectTypeNullable = isNullableType(objectType);
+    final isCastedTypeNullable = isNullableType(castedType);
+
+    // Only one case `Type? is Type` always valid assertion case.
+    return isObjectTypeNullable && !isCastedTypeNullable;
+  }
+
+  DartType? _castedTypeInObjectTypeHierarchy(
+    DartType objectType,
+    DartType castedType,
+  ) {
+    if (objectType.element == castedType.element) {
+      return objectType;
+    }
+
+    if (objectType is InterfaceType) {
+      return objectType.allSupertypes
+          .firstWhereOrNull((value) => value.element == castedType.element);
+    }
+
+    return null;
+  }
+
+  bool _isGenerics(DartType objectType, DartType castedType) {
```

Comment:
The naming doesn't really help understand what the method is doing.

Maybe something like this?
```suggestion
  bool _areGenericsWithSameTypeArgs(DartType objectType, DartType castedType) {
```